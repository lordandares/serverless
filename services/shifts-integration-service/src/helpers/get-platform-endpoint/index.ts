interface DeviceReceiver {
  deviceToken: string
  platform: string
}

export default async function getPlatformEndpoint(
  client: AWS.SNS,
  receiver: DeviceReceiver,
): Promise<AWS.SNS.CreateEndpointResponse | void> {
  const platformConfig = {
    android: {
      arn: process.env.AWS_SNS_ANDROID_ARN,
      service: process.env.AWS_SNS_ANDROID_SERVICE,
    },
    ios: {
      arn: process.env.AWS_SNS_IOS_ARN,
      service: process.env.AWS_SNS_IOS_SERVICE,
    },
  }

  const { deviceToken, platform } = receiver

  if (!deviceToken || !platform) return

  const config = platformConfig[platform]

  const params = {
    PlatformApplicationArn: config.arn,
    Token: deviceToken,
  }

  const endpoint = await client.createPlatformEndpoint(params).promise()

  return endpoint
}
