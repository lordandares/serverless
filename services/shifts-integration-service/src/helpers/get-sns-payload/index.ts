export default function getSnsPayload(params) {
  const { endpointArn, message, title, type: notificationType } = params

  return {
    Message: JSON.stringify({
      default: message,
      APNS: JSON.stringify({
        aps: {
          alert: message,
          'content-available': 1,
          sound: 'default',
        },
        notificationType,
      }),
      APNS_SANDBOX: JSON.stringify({
        aps: {
          alert: message,
          'content-available': 1,
          sound: 'default',
        },
        notificationType,
      }),
      GCM: JSON.stringify({
        data: {
          data: {
            notificationType,
          },
          message,
          title,
        },
      }),
    }),
    MessageStructure: 'json',
    TargetArn: endpointArn,
  }
}
