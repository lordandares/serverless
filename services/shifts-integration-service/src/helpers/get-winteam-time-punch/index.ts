import { errors, mongo, schemas } from '@lighthouse/serverless-common'
import { getOr, last } from 'lodash/fp'
import * as moment from 'moment-timezone'
import * as querystring from 'querystring'

// 2018-08-03 16:00:00.000-0400
const DATE_FORMAT = 'YYYY-MM-DD HH:mm:ss.SSSZZ'

interface TimePunchEvent extends TimePunch {
  event: string
}

interface TimePunch {
  result: {
    CallLogId: string
    HoursLogID?: string
    JobDescription: string
    JobID?: string
    JobNumber?: string
    PunchStatus: string
    PunchTime: string
    StatusReason?: string
  }
  shift: schemas.ShiftSchema
  user: schemas.UserSchema
}

interface TimePunchResponse {
  GeoCoordinate?: {
    Latitude: number
    Longitude: number
  }
  EmployeeId?: string
  EmployeeNumber?: string
  JobId?: string
  JobNumber?: string
  PunchTime: string
}

export async function getWinTeamTimePunch(
  timePunchEvent: TimePunchEvent,
): Promise<TimePunchResponse | void> {
  const { event, shift, user } = timePunchEvent
  const { application: applicationId, location: locationId } = shift

  if (!event) {
    throw new errors.ValidationError({
      message: 'Missing event from TimePunch',
    })
  }

  const employeeReference = getEmployeeReference(user)

  if (!employeeReference) {
    console.info('GetWinTeamTimePunch :: Missing Employee Reference', {
      shift,
      user,
    })

    return
  }

  const shiftEntity = getShiftEntity(shift, event)
  const shiftTimestamp = getOr(false, 'time', shiftEntity)

  if (!shiftTimestamp) {
    throw new errors.ValidationError({
      message: 'Missing timestamp from TimePunch',
      data: {
        shiftEntity,
        shiftTimestamp,
      },
    })
  }

  const jobReference = await getJobReference(
    shiftEntity,
    applicationId,
    locationId,
  )

  const punchTime = moment.utc(shiftTimestamp).format(DATE_FORMAT)

  const payload: TimePunchResponse = {
    ...employeeReference,
    ...jobReference,
    PunchTime: punchTime,
  }

  // Append GeoCoordinates if they exist
  const gpsCoordinates = getOr([], 'gps.geometry.coordinates', shiftEntity)

  if (gpsCoordinates && gpsCoordinates.length === 2) {
    payload.GeoCoordinate = {
      Latitude: gpsCoordinates[1],
      Longitude: gpsCoordinates[0],
    }
  }

  return payload
}

export async function getWinTeamMessages(
  timePunch: TimePunch,
): Promise<string | undefined> {
  const { result, shift, user } = timePunch

  const { JobNumber } = result

  const employeeReference = getEmployeeReference(user)
  if (!employeeReference) {
    console.info('GetWinTeamMessages :: Missing Employee Reference', {
      shift,
      user,
    })

    return
  }

  if (!JobNumber) {
    console.info('GetWinTeamMessages :: Missing Job Reference', {
      result,
      shift,
      user,
    })

    return
  }

  const punchTime = moment.utc(shift.start.time).format(DATE_FORMAT)

  const params = {
    ...employeeReference,
    JobNumber,
    PunchTime: punchTime,
  }

  return querystring.stringify(params)
}

export function getEmployeeReference(user: schemas.UserSchema): object | void {
  const employeeId = getOr(false, 'plugins.winteam.options.employeeId', user)
  const employeeNumber = getOr(
    false,
    'plugins.winteam.options.employeeNumber',
    user,
  )

  if (employeeId || employeeNumber) {
    return { EmployeeNumber: employeeId || employeeNumber }
  }
}

export async function getJobReference(
  shiftEntity: schemas.ShiftEventSchema,
  applicationId: string,
  locationId?: string,
): Promise<object | void> {
  const areaLocationId = getOr(false, 'area.location.id', shiftEntity)

  // If within an area location / legacy location, attempt to find the job number
  if (areaLocationId || locationId) {
    const isLegacy = !areaLocationId

    const collectionName = isLegacy ? 'locations' : 'areas'

    const queryLocationId = isLegacy ? locationId : areaLocationId
    const query = {
      _id: new mongo.ObjectId(queryLocationId),
      application: new mongo.ObjectId(applicationId),
    }

    const collection = await mongo.getCollection(collectionName)
    const location = await collection.findOne(query)

    if (location) {
      const jobId = getOr(false, 'plugins.winteam.options.jobId', location)
      const jobNumber = getOr(
        false,
        'plugins.winteam.options.jobNumber',
        location,
      )

      if (jobId || jobNumber) {
        return { JobNumber: jobId || jobNumber }
      }
    }
  }
}

export function getShiftEntity(
  shift: schemas.ShiftSchema,
  event: string,
): object | void {
  switch (event) {
    case 'shift-break-end':
      return last<schemas.ShiftBreakSchema>(shift.breaks)!.end
    case 'shift-end':
      return shift.end
    case 'shift-break-start':
      return last<schemas.ShiftBreakSchema>(shift.breaks)!.start
    case 'shift-start':
      return shift.start
    default:
      return {}
  }
}
