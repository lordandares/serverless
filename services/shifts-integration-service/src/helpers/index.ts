export { default as getPlatformEndpoint } from './get-platform-endpoint'
export { default as getSnsPayload } from './get-sns-payload'
export {
  getWinTeamMessages,
  getWinTeamTimePunch,
} from './get-winteam-time-punch'
export { default as notifyUser } from './notify-user'
export { default as parseWinTeamMessages } from './parse-winteam-messages'
export { default as sendMessage } from './send-message'
export {
  default as validateEndpointAttributes,
} from './validate-endpoint-attributes'
