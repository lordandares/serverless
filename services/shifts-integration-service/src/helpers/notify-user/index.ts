import { errors, schemas } from '@lighthouse/serverless-common'
import * as AWS from 'aws-sdk'
import { getOr, isEmpty } from 'lodash/fp'

import { getPlatformEndpoint, sendMessage } from '../'

interface NotifyUserParams {
  message: string
  user: schemas.UserSchema
}

export default async function notifyUser(
  params: NotifyUserParams,
): Promise<void> {
  const { message, user } = params

  if (!message || !user) {
    throw new errors.ValidationError({
      data: {
        message,
        user,
      },
      message: 'notifyUser :: Missing required params',
    })
  }

  const deviceReceivers = getOr([], 'speakerbox.deviceReceivers', user)

  if (isEmpty(deviceReceivers)) return

  const sns = new AWS.SNS()

  const endpoints: AWS.SNS.CreateEndpointResponse[] = await Promise.all(
    deviceReceivers.map(
      async receiver => await getPlatformEndpoint(sns, receiver),
    ),
  )

  await sendMessage(sns, endpoints, params)
}
