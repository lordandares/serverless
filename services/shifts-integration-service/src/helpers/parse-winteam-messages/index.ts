import { mongo } from '@lighthouse/serverless-common'
import { getOr } from 'lodash/fp'

export default function parseWinTeamMessages(winteamResponse) {
  const messages = getOr([], 'Messages', winteamResponse)

  return messages.map(message => {
    const {
      _id = new mongo.ObjectId(),
      Id,
      IsGlobal,
      Message,
      Response1,
      Response2,
      Response1Id,
      Response2Id,
      PunchActionID,
    } = message

    return {
      _id,
      isGlobal: !!IsGlobal,
      message: Message,
      messageId: Id,
      punchActionId: PunchActionID,
      response1: Response1,
      response1Id: Response1Id,
      response2: Response2,
      response2Id: Response2Id,
    }
  })
}
