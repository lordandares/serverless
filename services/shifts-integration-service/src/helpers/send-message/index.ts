import { getSnsPayload, validateEndpointAttributes } from '../'

interface SendMessageParams {
  message: string
  title: string
  type: string
}

export default async function sendMessage(
  client: AWS.SNS,
  endpoints: AWS.SNS.CreateEndpointResponse[],
  params: SendMessageParams,
): Promise<object[]> {
  const result = await Promise.all(
    endpoints.map(async endpoint => {
      const { EndpointArn: endpointArn } = endpoint

      await validateEndpointAttributes(client, endpointArn)

      const payload = getSnsPayload({ endpointArn, ...params })

      console.info('sending message', params)

      const result = await client.publish(payload).promise()

      console.info('sent message', result)
      return result
    }),
  )

  return result
}
