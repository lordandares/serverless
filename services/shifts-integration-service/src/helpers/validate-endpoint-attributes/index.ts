export default async function validateEndpointAttributes(
  client: AWS.SNS,
  arn: string,
): Promise<void> {
  const { Attributes } = await client
    .getEndpointAttributes({
      EndpointArn: arn,
    })
    .promise()

  const isEnabled = Attributes && Attributes.Enabled === 'true'

  if (isEnabled) return

  await client
    .setEndpointAttributes({
      EndpointArn: arn,
      Attributes: {
        ...Attributes,
        Enabled: 'true',
      },
    })
    .promise()
}
