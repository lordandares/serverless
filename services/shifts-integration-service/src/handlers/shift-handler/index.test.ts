const mockFindOne = jest.fn()
const mockGetCollection = jest.fn()

import { omit } from 'lodash/fp'
import { mongo } from '@lighthouse/serverless-common'

import * as AWSMock from 'aws-sdk-mock'
import * as AWS from 'aws-sdk'
import * as MockDate from 'mockdate'

AWSMock.setSDKInstance(AWS)

mongo.getCollection = mockGetCollection

import { createSnsEvent } from '../../../../../__test__/helpers'
import shiftHandler from './'

const getValidApplication = ({ winteamEnabled }) => ({
  _id: '000000000000000000000000',
  name: 'Valid Application',
  plugins: {
    winteam: {
      enabled: winteamEnabled,
    },
  },
})

const validEndShift = {
  _id: '111111111111111111111111',
  application: '000000000000000000000000',
  end: {
    time: '2019-06-01T07:00:00.000Z',
  },
  location: '222222222222222222222222',
  start: {
    time: '2019-06-01T05:00:00.000Z',
  },
  user: '333333333333333333333333',
}

const validStartShift = {
  _id: '111111111111111111111111',
  application: '000000000000000000000000',
  location: '222222222222222222222222',
  start: {
    time: '2019-06-01T05:00:00.000Z',
  },
  user: '333333333333333333333333',
}

const getValidUser = ({ winteamEnabled, ...options }) => ({
  _id: 'valid-user-id',
  email: 'valid-email-address',
  firstName: 'first-name',
  lastName: 'last-name',
  plugins: {
    winteam: {
      enabled: winteamEnabled,
      options,
    },
  },
  username: 'user-name',
})

let snsPublishMock = jest.fn()

describe('handlers:shift-handler', () => {
  beforeAll(() =>
    AWSMock.mock('StepFunctions', 'startExecution', snsPublishMock))

  beforeEach(() => {
    process.env.MONGODB_SECRET_ID = 'secretId'
    process.env.INTEGRATION_STATE_MACHINE_ARN =
      'arn:aws:states:us-east-1:0123456789012:stateMachine:state-machine-arn'

    MockDate.set('2000-01-01T00:00:00.000Z')
    jest.resetAllMocks()
  })

  afterEach(() => MockDate.reset())

  afterAll(() => AWSMock.restore())

  it('should error if the INTEGRATION_STATE_MACHINE_ARN is missing', async () => {
    expect.assertions(1)
    process.env.INTEGRATION_STATE_MACHINE_ARN = ''

    try {
      const payload = createSnsEvent({ body: '' })

      await shiftHandler(payload)
    } catch (err) {
      expect(err.message).toMatch(
        /ShiftError - INTEGRATION_STATE_MACHINE_ARN is missing/,
      )
    }
  })

  it('should error if the SNS message is empty', async () => {
    expect.assertions(1)

    try {
      const emptyPayload = createSnsEvent({ body: '' })

      await shiftHandler(emptyPayload)
    } catch (err) {
      expect(err.message).toMatch(/ShiftError - message could not be read/)
    }
  })

  it('should error if the SNS message is invalid', async () => {
    expect.assertions(1)

    try {
      const invalidPayload = createSnsEvent({ body: '{Invalid: json//' })

      await shiftHandler(invalidPayload)
    } catch (err) {
      expect(err.message).toMatch(/ShiftError - message could not be parsed/)
    }
  })

  it('should error if either the `application` or `_id` is not a valid ObjectId', async () => {
    expect.assertions(1)

    try {
      const invalidPayload = createSnsEvent({
        body: {
          application: ') && this.password.match(/*./)%00',
          shift: '111111111111111111111111',
        },
      })

      await shiftHandler(invalidPayload)
    } catch (err) {
      expect(err.message).toMatch(
        /Argument passed in must be a single String of 12 bytes or a string of 24 hex characters/,
      )
    }
  })

  it('should error if an application can not be retrieved', async () => {
    expect.assertions(1)

    mockGetCollection.mockResolvedValue({ findOne: mockFindOne })
    mockFindOne.mockResolvedValue(null)

    try {
      const missingApplication = createSnsEvent({
        body: {
          _id: '000000000000000000000000',
          application: '111111111111111111111111',
          event: 'start-shift',
        },
      })

      await shiftHandler(missingApplication)
    } catch (err) {
      expect(err.message).toMatch(/could not find application/)
    }
  })

  it('should error if a shift can not be retrieved', async () => {
    expect.assertions(1)

    mockGetCollection.mockResolvedValue({ findOne: mockFindOne })
    mockFindOne
      .mockResolvedValueOnce(getValidApplication({}))
      .mockResolvedValue(null)

    try {
      const invalidShift = createSnsEvent({
        body: {
          _id: '000000000000000000000000',
          application: '111111111111111111111111',
          event: 'start-shift',
        },
      })

      await shiftHandler(invalidShift)
    } catch (err) {
      expect(err.message).toMatch(/ShiftError - could not find shift/)
    }
  })

  it('should error if the user is missing', async () => {
    expect.assertions(1)

    const shiftWithoutUser = omit('user', validStartShift)

    mockGetCollection.mockResolvedValue({ findOne: mockFindOne })
    mockFindOne
      .mockResolvedValueOnce(getValidApplication({}))
      .mockResolvedValue(shiftWithoutUser)

    try {
      const invalidShift = createSnsEvent({
        body: {
          _id: '000000000000000000000000',
          application: '111111111111111111111111',
          event: 'start-shift',
        },
      })

      await shiftHandler(invalidShift)
    } catch (err) {
      expect(err.message).toMatch(/ShiftError - user missing from shift/)
    }
  })

  it('should error if the user cannot be retrieved', async () => {
    expect.assertions(1)

    mockGetCollection.mockResolvedValue({ findOne: mockFindOne })
    mockFindOne
      .mockResolvedValueOnce(getValidApplication({}))
      .mockResolvedValueOnce(validEndShift)
      .mockResolvedValue(null)

    try {
      const invalidShift = createSnsEvent({
        body: {
          _id: '000000000000000000000000',
          application: '111111111111111111111111',
          event: 'end-shift',
        },
      })

      await shiftHandler(invalidShift)
    } catch (err) {
      expect(err.message).toMatch(/could not find application user/)
    }
  })

  it('should set winteamEnabled=false if disabled at the application level', async () => {
    expect.assertions(3)

    snsPublishMock.mockResolvedValue('step-function-success')

    mockGetCollection.mockResolvedValue({ findOne: mockFindOne })
    mockFindOne
      .mockResolvedValueOnce(getValidApplication({ winteamEnabled: false }))
      .mockResolvedValueOnce(validStartShift)
      .mockResolvedValueOnce(
        getValidUser({
          employeeNumber: 'wt-1234',
          winteamEnabled: true,
        }),
      )

    const validMessage = createSnsEvent({
      body: {
        _id: '000000000000000000000000',
        application: '111111111111111111111111',
        event: 'start',
      },
    })

    const result = await shiftHandler(validMessage)

    expect(result).toEqual('step-function-success')

    expect(snsPublishMock).toHaveBeenCalledTimes(1)

    const input = JSON.parse(snsPublishMock.mock.calls[0][0].input)

    expect(input.plugins.winteamEnabled).toEqual(false)
  })

  it('should set winteamEnabled=false if disabled at the user level', async () => {
    expect.assertions(3)

    snsPublishMock.mockResolvedValue('step-function-success')

    mockGetCollection.mockResolvedValue({ findOne: mockFindOne })
    mockFindOne
      .mockResolvedValueOnce(getValidApplication({ winteamEnabled: true }))
      .mockResolvedValueOnce(validStartShift)
      .mockResolvedValueOnce(
        getValidUser({
          employeeNumber: 'wt-1234',
          winteamEnabled: false,
        }),
      )

    const validMessage = createSnsEvent({
      body: {
        _id: '000000000000000000000000',
        application: '111111111111111111111111',
        event: 'start',
      },
    })

    const result = await shiftHandler(validMessage)

    expect(result).toEqual('step-function-success')

    expect(snsPublishMock).toHaveBeenCalledTimes(1)

    const input = JSON.parse(snsPublishMock.mock.calls[0][0].input)

    expect(input.plugins.winteamEnabled).toEqual(false)
  })

  it('should successfully process a start shift', async () => {
    expect.assertions(4)

    snsPublishMock.mockResolvedValue('step-function-success')

    mockGetCollection.mockResolvedValue({ findOne: mockFindOne })
    mockFindOne
      .mockResolvedValueOnce(
        getValidApplication({
          winteamEnabled: true,
        }),
      )
      .mockResolvedValueOnce(validStartShift)
      .mockResolvedValueOnce(
        getValidUser({
          employeeNumber: 'wt-1234',
          winteamEnabled: true,
        }),
      )

    const validMessage = createSnsEvent({
      body: {
        _id: '000000000000000000000000',
        application: '111111111111111111111111',
        event: 'start',
      },
    })

    const result = await shiftHandler(validMessage)

    expect(result).toEqual('step-function-success')

    const snsMessage = snsPublishMock.mock.calls[0][0]
    const snsMessageInput = JSON.parse(snsMessage.input)

    expect(snsPublishMock).toHaveBeenCalledTimes(1)
    expect(snsMessage).toMatchSnapshot()

    expect(snsMessageInput.plugins.winteamEnabled).toEqual(true)
  })

  it('should successfully process an end shift', async () => {
    expect.assertions(4)

    snsPublishMock.mockResolvedValue('step-function-success')

    mockGetCollection.mockResolvedValue({ findOne: mockFindOne })
    mockFindOne
      .mockResolvedValueOnce(
        getValidApplication({
          winteamEnabled: true,
        }),
      )
      .mockResolvedValueOnce(validStartShift)
      .mockResolvedValueOnce(
        getValidUser({
          employeeNumber: 'wt-1234',
          winteamEnabled: true,
        }),
      )

    const validMessage = createSnsEvent({
      body: {
        _id: '000000000000000000000000',
        application: '111111111111111111111111',
        event: 'end',
      },
    })

    const result = await shiftHandler(validMessage)

    expect(result).toEqual('step-function-success')

    const snsMessage = snsPublishMock.mock.calls[0][0]
    const snsMessageInput = JSON.parse(snsMessage.input)

    expect(snsPublishMock).toHaveBeenCalledTimes(1)
    expect(snsMessage).toMatchSnapshot()

    expect(snsMessageInput.plugins.winteamEnabled).toEqual(true)
  })
})
