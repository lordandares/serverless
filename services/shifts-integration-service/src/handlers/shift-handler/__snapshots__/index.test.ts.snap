// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handlers:shift-handler should successfully process a start shift 1`] = `
Object {
  "input": "{\\"application\\":{\\"_id\\":\\"000000000000000000000000\\",\\"name\\":\\"Valid Application\\",\\"plugins\\":{\\"winteam\\":{\\"enabled\\":true}}},\\"event\\":\\"shift-start\\",\\"plugins\\":{\\"winteamEnabled\\":true},\\"shift\\":{\\"_id\\":\\"111111111111111111111111\\",\\"application\\":\\"000000000000000000000000\\",\\"location\\":\\"222222222222222222222222\\",\\"start\\":{\\"time\\":\\"2019-06-01T05:00:00.000Z\\"},\\"user\\":\\"333333333333333333333333\\"},\\"user\\":{\\"_id\\":\\"valid-user-id\\",\\"email\\":\\"valid-email-address\\",\\"firstName\\":\\"first-name\\",\\"lastName\\":\\"last-name\\",\\"plugins\\":{\\"winteam\\":{\\"enabled\\":true,\\"options\\":{\\"employeeNumber\\":\\"wt-1234\\"}}},\\"username\\":\\"user-name\\"}}",
  "name": "valid-user-id-shift-start-946684800000",
  "stateMachineArn": "arn:aws:states:us-east-1:0123456789012:stateMachine:state-machine-arn",
}
`;

exports[`handlers:shift-handler should successfully process an end shift 1`] = `
Object {
  "input": "{\\"application\\":{\\"_id\\":\\"000000000000000000000000\\",\\"name\\":\\"Valid Application\\",\\"plugins\\":{\\"winteam\\":{\\"enabled\\":true}}},\\"event\\":\\"shift-end\\",\\"plugins\\":{\\"winteamEnabled\\":true},\\"shift\\":{\\"_id\\":\\"111111111111111111111111\\",\\"application\\":\\"000000000000000000000000\\",\\"location\\":\\"222222222222222222222222\\",\\"start\\":{\\"time\\":\\"2019-06-01T05:00:00.000Z\\"},\\"user\\":\\"333333333333333333333333\\"},\\"user\\":{\\"_id\\":\\"valid-user-id\\",\\"email\\":\\"valid-email-address\\",\\"firstName\\":\\"first-name\\",\\"lastName\\":\\"last-name\\",\\"plugins\\":{\\"winteam\\":{\\"enabled\\":true,\\"options\\":{\\"employeeNumber\\":\\"wt-1234\\"}}},\\"username\\":\\"user-name\\"}}",
  "name": "valid-user-id-shift-end-946684800000",
  "stateMachineArn": "arn:aws:states:us-east-1:0123456789012:stateMachine:state-machine-arn",
}
`;
